src/core/Akka/Actor/Futures.cs

```diff
@@ -334,31 +378,39 @@ public override int GetHashCode()
         public static PromiseActorRef Apply(IActorRefProvider provider, TimeSpan timeout, object targetName,
             string messageClassName, IActorRef sender = null)
         {
-            sender = sender ?? ActorRefs.NoSender;
-            var result = new TaskCompletionSource<object>();
-            var a = new PromiseActorRef(provider, result, messageClassName);
-            var cancellationSource = new CancellationTokenSource();
-            cancellationSource.Token.Register(CancelAction, result);
-            cancellationSource.CancelAfter(timeout);
+            CancellationTokenSource cancellationSource = default;
 
-            //var scheduler = provider.Guardian.Underlying.System.Scheduler.Advanced;
-            //var c = new Cancelable(scheduler, timeout);
-            //scheduler.ScheduleOnce(timeout, () => result.TrySetResult(new Status.Failure(new AskTimeoutException(
-            //    string.Format("Ask timed out on [{0}] after [{1} ms]. Sender[{2}] sent message of type {3}.", targetName, timeout.TotalMilliseconds, sender, messageClassName)))),
-            //    c);
+            if (timeout != TimeSpan.Zero)
+            {
+                // avoid CTS + delegate allocation if timeouts aren't needed
+                cancellationSource = new CancellationTokenSource();
+                cancellationSource.CancelAfter(timeout);
+            }
 
-            result.Task.ContinueWith(r =>
+            var p = Apply(provider, messageClassName, cancellationSource?.Token ?? default);
+            
+            // need to dispose CTS afterwards
+            async Task ExecPromise()
```

Comment:
So the one thing that sucks here - have to have two `await`ed `Task`s for `GracefulStop` when a `TimeSpan` is used.

`GracefulStop` is not a performance-sensitive area so this ultimately doesn't matter - I just didn't like having to do it 😂